File: services/client/src/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"errors"
	"fmt"
	"net"
	"time"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/lottery"
	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

const (
	MSG_BATCH byte = 0x00
	MSG_FIN   byte = 0x01
	MSG_ACK   byte = 0x02
	MSG_HELLO byte = 0x03

	HEADER_SIZE      = 3
	MAX_PAYLOAD_SIZE = 65535
)

// PROTOCOLO
// => Header(3B) + payload(variable)
// Header:
//		msgType 1B
//		lenPayload 2B
//
// Payload de MSG_HELLO:
//		agencyId 1B
//
// Payload de MSG_BATCH:
//	lenBets 2B
//  N bets:
//		lenStrFirstName 2B
//		firstName (variable)
//		lenStrLastName 2B
//		lastName (variable)
//		Document 4B
//		birthday 4B
//		betNumber 4B

type Protocol struct {
	skt net.Conn
}

func NewProtocol(socket net.Conn) *Protocol {
	return &Protocol{skt: socket}
}

func (p *Protocol) Close() error {
	return p.skt.Close()
}

// SEND

func (p *Protocol) SendAck() error {
	return p.sendFrame(MSG_ACK, nil)
}

func (p *Protocol) SendHello(agencyId int) error {
	return p.sendFrame(MSG_HELLO, []byte{byte(agencyId)})
}

func (p *Protocol) SendNoMoreBets() error {
	return p.sendFrame(MSG_FIN, nil)
}

func (p *Protocol) SendBets(listOfBets []lottery.Bet) error {
	var payload []byte
	payload = binary.BigEndian.AppendUint16(payload, uint16(len(listOfBets)))

	for _, bet := range listOfBets {
		buffer, err := serializeBet(bet)
		if err != nil {
			return err
		}
		payload = append(payload, buffer...)
	}

	if err := p.sendFrame(MSG_BATCH, payload); err != nil {
		return err
	}
	return p.recvAck()
}

func serializeBet(bet lottery.Bet) ([]byte, error) {
	var buffer []byte
	buffer = appendString(buffer, bet.FirstName)
	buffer = appendString(buffer, bet.LastName)
	buffer = appendDocument(buffer, bet.Document)
	buffer, err := appendBirthdate(buffer, bet.Birthdate)
	if err != nil {
		return buffer, err
	}
	buffer = appendBetNumber(buffer, bet.Number)
	return buffer, nil
}

func (p *Protocol) sendFrame(msgType byte, payload []byte) error {
	if len(payload) > MAX_PAYLOAD_SIZE {
		return fmt.Errorf("payload too large: %d bytes", len(payload))
	}
	frame := []byte{msgType}
	frame = binary.BigEndian.AppendUint16(frame, uint16(len(payload)))
	frame = append(frame, payload...)
	return safe_socket.SendAll(p.skt, frame)
}

func appendString(buffer []byte, s string) []byte {
	lenBuf := make([]byte, 2)
	binary.BigEndian.PutUint16(lenBuf, uint16(len(s)))
	buffer = append(buffer, lenBuf...)
	return append(buffer, []byte(s)...)
}

func appendDocument(buffer []byte, document int) []byte {
	buf := make([]byte, 4)
	binary.BigEndian.PutUint32(buf, uint32(document))
	return append(buffer, buf...)
}

func appendBirthdate(buffer []byte, birthdate string) ([]byte, error) {
	date, err := time.Parse("2006-01-02", birthdate)
	if err != nil {
		return nil, err
	}
	buf := make([]byte, 4)
	binary.BigEndian.PutUint16(buf, uint16(date.Year()))
	buf[2] = byte(date.Month())
	buf[3] = byte(date.Day())
	buffer = append(buffer, buf...)
	return buffer, nil
}

func appendBetNumber(buffer []byte, betNumber int) []byte {
	buf := make([]byte, 4)
	binary.BigEndian.PutUint32(buf, uint32(betNumber))
	buffer = append(buffer, buf...)
	return buffer
}

func (p *Protocol) recvAck() error {
	msgType, _, err := p.recvFrame()
	if err != nil {
		return err
	}
	if msgType != MSG_ACK {
		return fmt.Errorf("expected ACK, got msgType %d", msgType)
	}
	return nil
}

// RECV

func (p *Protocol) RecvWinners() ([]lottery.Bet, bool, error) {
	msgType, payload, err := p.recvFrame()
	if err != nil {
		return nil, false, err
	}

	switch msgType {
	case MSG_FIN:
		return nil, false, nil
	case MSG_BATCH:
		winners, err := deserializeBets(payload)
		if err != nil {
			return nil, false, err
		}
		return winners, true, nil
	default:
		return nil, false, fmt.Errorf("unknown msgType: %d", msgType)
	}
}

func (p *Protocol) recvFrame() (byte, []byte, error) {
	header, err := safe_socket.RecvAll(p.skt, HEADER_SIZE)
	if err != nil {
		return 0, nil, err
	}
	payloadLen := int(binary.BigEndian.Uint16(header[1:3]))
	if payloadLen == 0 {
		return header[0], nil, nil
	}
	payload, err := safe_socket.RecvAll(p.skt, payloadLen)
	if err != nil {
		return 0, nil, err
	}
	return header[0], payload, nil
}

var errShortPayload = errors.New("payload shorter than expected")

func deserializeBets(payload []byte) ([]lottery.Bet, error) {
	if len(payload) < 2 {
		return nil, errShortPayload
	}

	countOfBets := int(binary.BigEndian.Uint16(payload[0:2]))
	offset := 2

	bets := make([]lottery.Bet, 0, countOfBets)
	for range countOfBets {
		bet, next, err := deserializeBet(payload, offset)
		if err != nil {
			return nil, err
		}
		bets = append(bets, bet)
		offset = next
	}

	if offset != len(payload) {
		return nil, fmt.Errorf("%d bytes left unparsed", len(payload)-offset)
	}

	return bets, nil
}

func deserializeBet(payload []byte, offset int) (lottery.Bet, int, error) {
	bet := lottery.Bet{}

	firstName, offset, err := deserializeString(payload, offset)
	if err != nil {
		return bet, offset, err
	}
	bet.FirstName = firstName
	lastName, offset, err := deserializeString(payload, offset)
	if err != nil {
		return bet, offset, err
	}
	bet.LastName = lastName

	if offset+12 > len(payload) {
		return bet, offset, errShortPayload
	}
	buffer := payload[offset : offset+12]
	bet.Document = int(binary.BigEndian.Uint32(buffer[:4]))
	bet.Birthdate = deserializeBirthdate(buffer[4:8])
	bet.Number = int(binary.BigEndian.Uint32(buffer[8:]))
	offset += 12

	return bet, offset, nil
}

func deserializeString(payload []byte, offset int) (string, int, error) {
	if offset+2 > len(payload) {
		return "", offset, errShortPayload
	}
	length := int(binary.BigEndian.Uint16(payload[offset : offset+2]))
	offset += 2
	if offset+length > len(payload) {
		return "", offset, errShortPayload
	}
	return string(payload[offset : offset+length]), offset + length, nil
}

func deserializeBirthdate(buffer []byte) string {
	year := binary.BigEndian.Uint16(buffer[:2])
	month := buffer[2]
	day := buffer[3]
	return fmt.Sprintf("%04d-%02d-%02d", year, month, day)
}

File: services/server/src/protocol/protocol.py
import safe_socket
from lottery.bet import Bet

MSG_BATCH = 0x00
MSG_FIN = 0x01
MSG_ACK = 0x02
MSG_HELLO = 0x03

HEADER_SIZE = 3
MAX_PAYLOAD_SIZE = 65535

SHORT_PAYLOAD = "payload shorter than expected"

class Protocol:
    def __init__(self, skt):
        self.skt = skt

    def close(self):
        return self.skt.close()

    # SEND
    def send_bets(self, bets):
        payload = len(bets).to_bytes(2, "big")
        for bet in bets:
            payload += self._serialize_bet(bet)
        self._send_frame(MSG_BATCH, payload)

    def _send_frame(self, msg_type, payload=b""):
        if len(payload) > MAX_PAYLOAD_SIZE:
            raise ValueError(f"payload too large: {len(payload)} bytes")
        frame = msg_type.to_bytes(1, "big")
        frame += len(payload).to_bytes(2, "big")
        frame += payload
        safe_socket.send_all(self.skt, frame)

    def _serialize_bet(self, bet):
        buf = self._serialize_string(bet.first_name)
        buf += self._serialize_string(bet.last_name)
        buf += bet.document.to_bytes(4, "big")
        buf += self._serialize_birthdate(bet.birthdate)
        buf += bet.number.to_bytes(4, "big")
        return buf

    def _serialize_string(self, s):
        encoded = s.encode("utf-8")
        return len(encoded).to_bytes(2, "big") + encoded

    def _serialize_birthdate(self, birthdate):
        year, month, day = map(int, birthdate.split("-"))
        return year.to_bytes(2, "big") + month.to_bytes(1, "big") + day.to_bytes(1, "big")
    
    def recv_ack(self):
        msg_type, _ = self._recv_frame()
        if msg_type != MSG_ACK:
            raise ValueError(f"expected ACK, got msg_type {msg_type}")

    def send_ack(self):
        self._send_frame(MSG_ACK)

    def send_no_more_bets(self):
        self._send_frame(MSG_FIN)


    # RECV

    def recv_hello(self):
        msg_type, payload = self._recv_frame()
        if msg_type != MSG_HELLO:
            raise ValueError(f"expected MSG_HELLO, got msgType {msg_type}")
        if len(payload) != 1:
            raise ValueError(SHORT_PAYLOAD)
        return payload[0]

    def recv_bets(self, agency_id): # devuelve (bets, is_fin)
        msg_type, payload = self._recv_frame()
        if msg_type == MSG_FIN:
            return [], True
        elif msg_type == MSG_BATCH:
            return self._deserialize_bets(payload, agency_id), False
        else:
            raise ValueError(f"unknown msgType: {msg_type}")

    def _recv_frame(self):
        header = safe_socket.recv_all(self.skt, HEADER_SIZE)
        if not header:
            raise ConnectionError("closed socket")
        payload_len = int.from_bytes(header[1:3], "big")
        payload = safe_socket.recv_all(self.skt, payload_len) if payload_len else b""
        return header[0], payload

    def _deserialize_bets(self, payload, agency_id):
        if len(payload) < 2:
            raise ValueError(SHORT_PAYLOAD)

        count_of_bets = int.from_bytes(payload[0:2], "big")
        offset = 2

        bets = []
        for _ in range(count_of_bets):
            bet, offset = self._deserialize_bet(payload, offset, agency_id)
            bets.append(bet)

        if offset != len(payload):
            raise ValueError(f"{len(payload) - offset} bytes left unparsed")
        return bets

    def _deserialize_bet(self, payload, offset, agency_id):
        first_name, offset = self._deserialize_string(payload, offset)
        last_name, offset = self._deserialize_string(payload, offset)

        if offset + 12 > len(payload):
            raise ValueError(SHORT_PAYLOAD)
        document = int.from_bytes(payload[offset:offset + 4], "big")
        birthdate = self._deserialize_birthdate(payload[offset + 4:offset + 8])
        number = int.from_bytes(payload[offset + 8:offset + 12], "big")
        offset += 12

        return Bet(agency_id, first_name, last_name, document, birthdate, number), offset

    def _deserialize_string(self, payload, offset):
        if offset + 2 > len(payload):
            raise ValueError(SHORT_PAYLOAD)
        length = int.from_bytes(payload[offset:offset + 2], "big")
        offset += 2
        if offset + length > len(payload):
            raise ValueError(SHORT_PAYLOAD)
        return payload[offset:offset + length].decode("utf-8"), offset + length

    def _deserialize_birthdate(self, buffer):
        year = int.from_bytes(buffer[:2], "big")
        return f"{year:04d}-{buffer[2]:02d}-{buffer[3]:02d}"